src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs

```diff
@@ -0,0 +1,64 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Diagnostics
+{
+    /// <summary>
+    /// RandomNumberGenerator implementation is the 64-bit random number generator based on the Xorshift algorithm (known as shift-register generators).
+    /// </summary>
+    internal class RandomNumberGenerator
+    {
+        [ThreadStatic] private static RandomNumberGenerator? t_random;
+
+        private ulong _s0, _s1, _s2, _s3;
+
+        public static RandomNumberGenerator Current
+        {
+            get
+            {
+                if (t_random == null)
+                {
+                    t_random = new RandomNumberGenerator();
+                }
+                return t_random;
+            }
+        }
+
+#if ALLOW_PARTIALLY_TRUSTED_CALLERS
+        [System.Security.SecuritySafeCriticalAttribute]
+#endif
+        public unsafe RandomNumberGenerator()
+        {
+            do
+            {
+                Guid g1 = Guid.NewGuid();
```

Comment:
@tarekgh one observation: some bits are fixed on the `ulong` captured from the `Guid`. I don't think this is an actual problem since this already has enough bits of entropy but if you want the full 64-bits of entropy you should replace the byte with the fixed bits with a byte without fixed bits from the second-half of the `Guid` (which also has some fixed bits).

Reply:
Thanks @pjanotti. That make sense. I have submitted a commit randomizing these bits. please have a look and let me know if you have any comments.

Reply:
LGTM @tarekgh, thanks for this change.

Reply:
ops `*(g1p + 1)` also has 2 fixed bits...

Reply:
I have submitted one more commit to address that. Thanks again.